### PR TITLE
SOLR-15392 Tracing span operation name and tags

### PR DESCRIPTION
**Pull request by dsmiley**
https://issues.apache.org/jira/browse/SOLR-15392

### CODE REVIEW

**Comment by dsmiley** [comment]
I think this is ready for review/merge.  I'll may wait for more internal use where I work before merging.  The V2 API doesn't result in ideal span operation names (looks like `post:/c/collection1` instead of `create:/c/{collection}`, however, and I think that should come in a separate PR that follows some of the work in #81 which will likely affect the approach on how to solve it.

File: solr/core/src/java/org/apache/solr/request/SolrQueryRequestBase.java [hunk 1]
```diff
@@ -144,19 +146,31 @@ public IndexSchema getSchema() {
...
-    }
     final HttpSolrCall call = getHttpSolrCall();
     if (call != null) {
       final Tracer tracer = (Tracer) call.getReq().getAttribute(Tracer.class.getName());
       if (tracer != null) {
         return tracer;
       }
     }
+    if (core != null) {
+      return core.getCoreContainer().getTracer();
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/request/SolrQueryRequestBase.java:157]: If _hypothetically_ (maybe in a test or some interesting custom plugins) have a Tracer on the request that is different than a more static/global tracer, then we should prefer the request tracer.  Very hypothetical!

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java [hunk 2]
```diff
@@ -499,6 +495,7 @@ public void doFilter(ServletRequest _request, ServletResponse _response, FilterC
         if (log.isDebugEnabled()) {
           log.debug("User principal: {}", request.getUserPrincipal());
         }
+        span.setTag(Tags.DB_USER, String.valueOf(request.getUserPrincipal()));
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:498]: This line is not reached if there is no auth enabled, and thus there will be no `db.user` tag set.

File: solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java [hunk 3]
```diff
@@ -171,23 +173,50 @@ public SolrQueryRequest parse( SolrCore core, String path, HttpServletRequest re
...
+    String pathNormalized = RequestHandlers.normalize(path);
+    sreq.getContext().put(PATH, pathNormalized);
     sreq.getContext().put("httpMethod", req.getMethod());
 
     if(addHttpRequestToContext) {
       sreq.getContext().put("httpRequest", req);
     }
+
+    // add span info
+    TraceUtils.ifNotNoop(
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java]: In my most recent commit, I moved the logic to a method on HttpSolrCall and thus it becomes a method reference.

File: solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java [hunk 4]
```diff
@@ -549,7 +546,9 @@ public InputStream getStream() throws IOException {
     public SolrParams parseParamsAndFillStreams( 
         final HttpServletRequest req, ArrayList<ContentStream> streams ) throws Exception
     {
-      streams.add( new HttpRequestContentStream( req ) );
+      if (!req.getMethod().equals("GET")) { // or other conditions? Check headers?
+        streams.add(new HttpRequestContentStream(req));
+      }
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java:551]: @noblepaul  when doing a GET request for V2, I found it to be a problem that we add this stream that is empty, resulting in an exception at some point about a blank Content-Type (as there is no content in the first place).  StandardRequestParser.parseParamsAndFillStreams checks !isPost and isV2 and then calls this raw impl.
I could probably be more thorough here in detecting the presence of a stream.  Just because it's not a GET doesn't mean there is a stream!  Weirdly, req.getInputStream can return non-null for a GET which I observed in a test I added in this PR.

**Review by dsmiley — COMMENTED** [review]
@noblepaul I at-mentioned you on some of this due to your involvement in the V2 API.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java [hunk 5]
```diff
@@ -360,6 +366,85 @@ protected void execute(SolrQueryResponse rsp) {
...
+      if (cmds.size() == 1) {
+        verb = cmds.get(0).name;
+      }
+    }
+
+    span.setOperationName(verb + ":" + path);
+
+    if (req.getQueryString() != null) {
+      span.setTag("params", req.getQueryString());
+    }
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/api/V2HttpCall.java:410]: I did a change here that is debatable.  Previously, this was the SolrParams toString but is now the non-decoded / raw query string (thus has percent-encoding).  But I think I want to change this back.  This is all guarded by no-op, and furthermore we could modify SolrQueryRequestBase to cache the toString of the original params if we want.  That method is also called by SolrCore.preDecorateResponse for logging.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java [hunk 6]
```diff
@@ -360,6 +366,85 @@ protected void execute(SolrQueryResponse rsp) {
...
+      List<CommandOperation> cmds = solrReq.getCommands(false);
+      if (cmds.size() == 1) {
+        verb = cmds.get(0).name;
+      }
+    }
+
+    span.setOperationName(verb + ":" + path);
+
+    if (req.getQueryString() != null) {
+      span.setTag("params", req.getQueryString());
```
Comment (haythemkh) [inline solr/core/src/java/org/apache/solr/api/V2HttpCall.java:409]: would "params" need to be added as a constant? for example `Tags.PARAMS`

File: solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java (hunk 4 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java:551]: I should enhance this condition to look for `Content-Length` or `Transfer-Encoding` headers, which the HTTP RFC says is how you know if there is a message body.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 6 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:409]: I don't think there is value in a constant for this use-case (basically a form of logging).  If hypothetically this was parsed within the codebase then it would make sense to me as it would also link the dependency between read/write.

Also, I think this should be namespaced... like "http.params".  The names seem somewhat loose; "http." is not reserved, so I think it's okay to add it there.

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (haythemkh) [reply solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:498]: Would it be more interesting to specify that the user is not authenticated via a value of None for example instead of bypassing this setting?

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java [hunk 7]
```diff
@@ -603,6 +608,44 @@ public Action call() throws IOException {
...
+    }
+    if (coreOrColName != null) {
+      span.setTag(Tags.DB_INSTANCE, coreOrColName);
+    }
+
+    // Set operation name.
+    String path = getPath();
+    if (coreOrColName != null) {
+      // prefix path by core or collection name
+      if (getCore() != null && getCore().getName().equals(coreOrColName)) {
```
Comment (haythemkh) [inline solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:632]: why not nest this block in the block above (line 621)?

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java [hunk 8]
```diff
@@ -603,6 +608,44 @@ public Action call() throws IOException {
...
+        path = "/{core}" + path;
+      } else {
+        path = "/{collection}" + path;
+      }
+    }
+    String verb =
+        getQueryParams()
+            .get(CoreAdminParams.ACTION, req.getMethod())
+            .toLowerCase(Locale.ROOT);
+    span.setOperationName(verb + ":" + path);
```
Comment (haythemkh) [inline solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:642]: is it possible that verb setting is null? is it tolerable?

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java [hunk 9]
```diff
@@ -533,10 +533,23 @@ public void doFilter(ServletRequest _request, ServletResponse _response, FilterC
...
     }
   }
-  
+
+  protected Span buildSpan(HttpServletRequest request, Tracer tracer) {
+    return tracer
+        .buildSpan(
+            "http.request") // will be changed later
+        .asChildOf(tracer.extract(Format.Builtin.HTTP_HEADERS, new HttpServletCarrier(request)))
+        .withTag(Tags.SPAN_KIND, Tags.SPAN_KIND_SERVER)
```
Comment (haythemkh) [inline solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:546]: do we need to add port and host as tags to keep alignment with the deleted code? or is it not necessary as it is included in the url? or were they deleted for some reason?

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java [hunk 10]
```diff
@@ -533,10 +533,23 @@ public void doFilter(ServletRequest _request, ServletResponse _response, FilterC
...
+        .withTag(Tags.HTTP_METHOD, request.getMethod())
```
Comment (haythemkh) [inline solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:547]: do we need to add port and host as tags to keep alignment with the deleted code? or is it not necessary as it is included in the url? or were they deleted for some reason?

File: solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java [hunk 11]
```diff
@@ -171,23 +169,22 @@ public SolrQueryRequest parse( SolrCore core, String path, HttpServletRequest re
...
-    Span span = (Span) req.getAttribute(Span.class.getName()); // not null but maybe in some tests?
-    if (span != null && !(span instanceof NoopSpan)) {
-      span.setTag("params", params.toString());
-    }
     SolrQueryRequest sreq = buildRequestFrom(core, params, streams, getRequestTimer(req), req);
 
     // Handlers and login will want to know the path. If it contains a ':'
     // the handler could use it for RESTful URLs
-    sreq.getContext().put(PATH, RequestHandlers.normalize(path));
+    String pathNormalized = RequestHandlers.normalize(path);
```
Comment (haythemkh) [inline solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java:176]: why did you need to split the above row in half?

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java (hunk 7 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:632]: _Maybe_... but I find it clearer to separate the two separate objectives rather than intertwine them.

File: solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java (hunk 8 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/HttpSolrCall.java:642]: Impossible.  See the "get" that takes 2 args.  The 2nd param there is the default, which is the "method" (HTTP) of the request, which can't be null.

File: solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java (hunk 11 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java:176]: The code placement of some of my changes moved in the course of multiple commits on this PR, but didn't leave things identical to how things were originally.  I will undo this.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 6 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:409]: In a commit soon I moved this to SolrDispatchFilter as http.params where there is no duplication and also where it's aligned with the "http" tags.

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java (hunk 10 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:547]: I think you are observing from the diff that the span operation name used to include the host & port.  Those are poor choices for a span operation name, and this PR expressly sets out to improve this.  They weren't in dedicated tags, although HTTP_URL includes them if we're curious.

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:500]: If Auth is enabled but the user isn't authenticated, getUserPrincipal() will return null and then String.valueOf will yield "null".  I don't think we should add this header if there is no auth because the header is then inapplicable.

**Comment by dsmiley** [comment]
I separated out https://issues.apache.org/jira/browse/SOLR-15418 #115 for the ContentStream stuff that is unrelated (but tickled by immature V2 codepaths).

File: solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java [hunk 12]
```diff
@@ -202,7 +202,7 @@ public void handleRequest(SolrQueryRequest req, SolrQueryResponse rsp) {
       }
     }
     Timer.Context timer = requestTimes.time();
-    @SuppressWarnings("resource")
+    //@SuppressWarnings("resource")
```
Comment (haythemkh) [inline solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java:205]: curious: why not deleting full line? :)

Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/handler/RequestHandlerBase.java:205]: A mistake; I'll remove -- thanks.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java [hunk 13]
```diff
@@ -360,6 +366,85 @@ protected void execute(SolrQueryResponse rsp) {
...
+
+    // Get the templatize-ed path
+    String path;
+    if (api instanceof AnnotatedApi) {
+      // ideal scenario; eventually everything might be AnnotatedApi?
+      var aapi = (AnnotatedApi) api;
+      path = aapi.getEndPoint().path()[0]; // consider first to be primary
+    } else {
+      path = computeEndpointPath();
+      // TODO consider getValidators, looking for command & path?
```
[the remark this answers is not in the document]
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:391]: this is correct @dsmiley

[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:391]: @noblepaul I'm looking at it now and I don't see how getValidators() can tell me what the template path is for the current request (the equivalent of `EndPoint.path()`).  Can you advise?

The `computeEndpointPath() algorithm seems pretty reasonable to me, and not slow.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java [hunk 14]
```diff
@@ -360,6 +366,80 @@ protected void execute(SolrQueryResponse rsp) {
...
+    }
+
+    String verb = req.getMethod().toLowerCase(Locale.ROOT);
+    // if this api has commands ...
+    final Map<String, JsonSchemaValidator> validators = getValidators(); // should be cached
+    if (validators != null && validators.isEmpty() == false && solrReq != null) {
+      // does this request have one command?
+      List<CommandOperation> cmds = solrReq.getCommands(false);
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/api/V2HttpCall.java:399]: @noblepaul I'm unsure if `getCommands(false)` should be that or `true` -- a "validateInput" argument.  I was thinking not validate to save CPU for this use case but getCommands caches its result, and thus I think future calls to getCommands(true) will return cached results of non-validated commands.  And that will happen typically.  I'll change this back to `true`.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java [hunk 15]
```diff
@@ -360,6 +366,80 @@ protected void execute(SolrQueryResponse rsp) {
...
+      path = ((AnnotatedApi) api).getEndPoint().path()[0]; // consider first to be primary
+    } else {
+      path = computeEndpointPath();
+      // TODO consider getValidators, looking for command & path?
+    }
+
+    String verb = req.getMethod().toLowerCase(Locale.ROOT);
+    // if this api has commands ...
+    final Map<String, JsonSchemaValidator> validators = getValidators(); // should be cached
+    if (validators != null && validators.isEmpty() == false && solrReq != null) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/api/V2HttpCall.java:397]: This is somewhat of a hack for performance.  Ideally, solrReq.getCommands below would eagerly grab the validators to see none and then return an empty list before bothering with grabbing the streams.  I could do a separate PR for that. CC @noblepaul

File: solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java (hunk 4 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java]: Please enhance this and I shall review it

Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/servlet/SolrRequestParsers.java]: See #134 -- I split this part out to its own.  I requested your review there already.

**Comment by dsmiley** [comment]
Synced with main thus incorporating #134 .  Any final thoughts @noblepaul ?

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java [hunk 16]
```diff
@@ -360,6 +366,80 @@ protected void execute(SolrQueryResponse rsp) {
...
+    String path;
+    if (api instanceof AnnotatedApi) {
+      // ideal scenario; eventually everything might be AnnotatedApi?
+      path = ((AnnotatedApi) api).getEndPoint().path()[0]; // consider first to be primary
+    } else {
+      path = computeEndpointPath();
+      // TODO consider getValidators, looking for command & path?
+    }
+
+    String verb = req.getMethod().toLowerCase(Locale.ROOT);
```
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/api/V2HttpCall.java:394]: why use to `toLowerCase(Locale.ROOT);`

it's expensive. it's always going to be in ALL CAPS

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 15 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:397]: The most common usecase will not have any commands. So, avoid getting the validators and the verb is absent

**Review by noblepaul — CHANGES_REQUESTED** [review]

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 16 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:394]: It's common in tracing (from my limited experience) to use lower-case, so I'm trying to be consistent here.  Also I very much disagree on the relative cost of this compared to the limited cases where this is activated.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 14 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java]: I would use `false`

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 15 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:397]: I just wanted this part of the code to be as cheap as posisble

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 16 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java:394]: is this used rarely or in every request?

Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java]: If someone opts-in to tracing, then every request is traced but not necessarily "reported" to a tracing server (maybe 1% depending on configuration).  Unfortunately even a non-reported span will be decorated with this metadata that is ultimately not used.  Ideally the tracing impl chosen can return a Span implementing NoopSpan which will avoid this logic here but neither Jaeger nor Brave OpenTracing adapters do this :-/  I don't know about others.  I reported this enhancement idea to Brave.

Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java]: BTW when I said 

> the relative cost of this compared to the limited cases where this is activated.

I not only meant opt-in to tracing but in the big picture of all that goes on to field a request into Solr to actually do something.  Lowercasing a string is inconsequential.

**Comment by dsmiley** [comment]
Are we good here @noblepaul ?  You set "change requested" review status.

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java]: ` Api.getSpec()._get("url", "paths")[0]`

**Comment by noblepaul** [comment]
LGTM @dsmiley

File: solr/core/src/java/org/apache/solr/api/V2HttpCall.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/api/V2HttpCall.java]: Thanks!
I tried this:
```
      @SuppressWarnings("unchecked")
      List<String> paths = (List<String>) api.getSpec()._get(List.of("url", "paths"), null);
      if (paths != null && !paths.isEmpty()) {
        // TODO use "paths[0]" above but would rather see a faster implementation first
        path = paths.get(0);
      }
```
And it worked for some endpoints but not for collection/core handlers which falsely returned "/update" instead of "/collections/{collection}/update".  And when /json was tack'ed on, it wasn't reflected there either but I suppose that could be minor.

So I'll stick with the current code (which you approved today).
